src/OpenTelemetry.Contrib.Instrumentation.AWSLambda/Implementation/AWSLambdaWrapper.cs

```diff
@@ -27,162 +27,219 @@ namespace OpenTelemetry.Contrib.Instrumentation.AWSLambda.Implementation
+        /// If parent is not extracted from incoming request then X-Ray propagation is used to extract one
+        /// unless X-Ray propagation is disabled in the configuration for this wrapper.
+        /// </param>
         /// <returns>Task.</returns>
-        public static async Task Trace<TInput>(TracerProvider tracerProvider, Func<TInput, ILambdaContext, Task> lambdaHandler, TInput input, ILambdaContext context)
-        {
-            await Intercept(tracerProvider, () => lambdaHandler(input, context), context);
-        }
-
-        /// <summary>
-        /// Tracing wrapper for async Lambda handler.
-        /// </summary>
-        /// <typeparam name="TInput">Input.</typeparam>
-        /// <typeparam name="TResult">Output result.</typeparam>
-        /// <param name="tracerProvider">TracerProvider passed in.</param>
-        /// <param name="lambdaHandler">Lambda handler function passed in.</param>
-        /// <param name="input">Instance of input.</param>
-        /// <param name="context">Instance of lambda context.</param>
-        /// <returns>Task of result.</returns>
-        public static async Task<TResult> Trace<TInput, TResult>(TracerProvider tracerProvider, Func<TInput, ILambdaContext, Task<TResult>> lambdaHandler, TInput input, ILambdaContext context)
+        public static Task Trace(
+            TracerProvider tracerProvider,
+            Func<ILambdaContext, Task> lambdaHandler,
+            ILambdaContext context,
+            ActivityContext parentContext = default)
         {
-            return await Intercept(tracerProvider, () => lambdaHandler(input, context), context);
+            Func<Task> action = async () => await lambdaHandler(context);
+            return TraceInternalAsync<object>(tracerProvider, action, null, context, parentContext);
         }
 
-        private static TResult Intercept<TResult>(TracerProvider tracerProvider, Func<TResult> method, ILambdaContext context = null)
+        internal static Activity OnFunctionStart<TInput>(TInput input, ILambdaContext context, ActivityContext parentContext = default)
         {
-            var lambdaActivity = OnFunctionStart(context);
-            try
+            if (parentContext == default)
             {
-                return method();
+                parentContext = AWSLambdaUtils.ExtractParentContext(input);
+                if (parentContext == default && !DisableAwsXRayContextExtraction)
+                {
+                    parentContext = AWSLambdaUtils.GetXRayParentContext();
+                }
             }
-            catch (Exception ex)
-            {
-                OnException(lambdaActivity, ex);
 
-                throw;
-            }
-            finally
-            {
-                OnFunctionStop(lambdaActivity, tracerProvider);
-            }
+            var tags = AWSLambdaUtils.GetFunctionTags(input, context);
+            var activityName = AWSLambdaUtils.GetFunctionName(context) ?? "<Unknown>";
```

Comment:
```suggestion
            var activityName = AWSLambdaUtils.GetFunctionName(context) ?? "AWS Lambda Invoke";
```

(although that should hopefully be never happen in practice)